File: problems/C++/65. Valid Number.cpp
class Solution {
public:
    bool isNumber(string s) {
        bool dig = false;
        bool dot = false;
        bool exp = false;
        
        for (int i = 0; i < s.length(); ++i) {
            char c = s[i];
            
            if (isdigit(c)) {
                dig = true;
            } else if (c == '+' || c == '-') {
                if (i > 0 && tolower(s[i-1]) != 'e') {
                    return false;
                }
            } else if (c == '.') {
                if (dot || exp) return false;
                dot = true;
            } else if (tolower(c) == 'e') {
                if (exp || !dig) return false;
                exp = true;
                dig = false;
            } else {
                return false;
            }
        }
        return dig;
    }
};

File: problems/Python/65. Valid Number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        dig, dot, exp = False, False, False
        
        for i, c in enumerate(s):
            if c.isdigit():
                dig = True
            elif c in ['+', '-']:
                if i > 0 and s[i-1].lower() != 'e':
                    return False
            elif c == '.':
                if dot or exp:
                    return False
                dot = True
            elif c.lower() == 'e':
                if exp or not dig:
                    return False
                exp = True
                dig = False
            else:
                return False
                
        return dig